### PR TITLE
GAIL-PPO

### PR DESCRIPTION
**Pull request by lbertge**
Implements PPO as the generator policy in GAIL. 
I noticed from the argparse methods, that PPO was supposed to be a choice, so I wondered if it was possible to support it. After some work, I have been able to get it to converge consistently, with similar performance characteristics as TRPO. For this reason, I believe it's a valuable addition to this repository. 

I've shifted some code around to make it a little less repetitive, but shouldn't change any method signatures or parameters. I did make some slight changes to the argparser, which in retrospect may be breaking, so I'm willing to go back and redo those.

The current GAIL codebase runs on TF 1.14. 

Initial Benchmarks:
image

image

GAIL-PPO has also been configured to support MPI. The graphs are not as good-looking, but I hope it helps support my case:
image

Some additional thoughts:
I have tried but have not had much success with integrating PPO2's implementation with GAIL. A bit more investigation led me to believe that PPO1 doesn't clip its [value loss](https://github.com/openai/baselines/blob/master/baselines/ppo1/pposgd_simple.py#L114), but PPO2 [seems to](https://github.com/openai/baselines/blob/master/baselines/ppo2/model.py#L75). I note some other suspected changes, but I'd like to know if this is worth investigating before opening up a new discussion. Thank you, I'd appreciate your time.

### CODE REVIEW

**Comment by pzhokhov** [comment]
Hi @lbertge Thanks for your PR, this looks like a neat addition! Unfortunately, at the moment we are not accepting new features to baselines (mostly due to lack of the resources to maintain existing codebase, let alone a bigger one). I would actually recommend re-submitting a PR to the stable-baselines project (https://github.com/hill-a/stable-baselines)

**Comment by Kailiangdong** [comment]
Thank you. I am currently looking for baselines/gail using ppo. Finally , I'd like to have GAIL with ppo2 and wassersteinGAN.
Can you tell more about intergration of ppo2? I think it is worth. We can open a new discussion about that.      Thx.

File: baselines/gail/run_mujoco.py [hunk 1]
```diff
@@ -129,7 +142,7 @@ def train(env, seed, policy_fn, reward_giver, dataset, algo,
         pretrained_weight = behavior_clone.learn(env, policy_fn, dataset,
                                                  max_iters=BC_max_iter)
 
-    if algo == 'trpo':
+    if alg == 'trpo':
```
Comment (Kailiangdong) [inline baselines/gail/run_mujoco.py]: Why you here still write ’trpo'? Thx

File: baselines/gail/run_mujoco.py [hunk 2]
```diff
@@ -36,13 +41,13 @@ def argsparser():
...
     # Optimization Configuration
-    parser.add_argument('--g_step', help='number of steps to train policy in each epoch', type=int, default=3)
+    parser.add_argument('--g_step', help='number of steps to train policy in each epoch', type=int, default=1)
     parser.add_argument('--d_step', help='number of steps to train discriminator in each epoch', type=int, default=1)
     # Network Configuration (Using MLP Policy)
     parser.add_argument('--policy_hidden_size', type=int, default=100)
     parser.add_argument('--adversary_hidden_size', type=int, default=100)
     # Algorithms Configuration
-    parser.add_argument('--algo', type=str, choices=['trpo', 'ppo'], default='trpo')
+    parser.add_argument('--alg', type=str, choices=['trpo', 'ppo'], default='trpo')
```
Comment (Kailiangdong) [inline baselines/gail/run_mujoco.py]: default should be change to "ppo" right?

**Comment by lbertge** [comment]
Hi all,

thank you for your responses. @pzhokhov, I will do my best to look at stable-baselines.

@Kailiangdong, you're welcome to open an issue on my fork if you want to discuss it further. You should be able to clone the repo and run it on TF 1.14.

**Comment by Kailiangdong** [comment]
@lbertge  Thank you. It seems that your repository don't open the "Issue".
I have a question, why you update G only once in ppo. But the original one is to update G for three times.

**Pull request closed without merging**